src/event/HandlerList.php

```diff
@@ -129,24 +128,23 @@ public function getListenerList() : array{
 			$handlerLists[] = $currentList;
 		}
 
-		$listenersByPriority = [];
-		$asyncListenersByPriority = [];
+		$listeners = [];
+		$asyncListeners = [];
+		$exclusiveAsyncListeners = [];
 		foreach($handlerLists as $currentList){
-			foreach($currentList->handlerSlots as $priority => $listeners){
-				$syncListeners = array_filter($listeners, static function(RegisteredListener $listener) : bool{ return !($listener instanceof RegisteredAsyncListener); });
-				$listenersByPriority[$priority] = array_merge($listenersByPriority[$priority] ?? [], $syncListeners);
-
-				$asyncListeners = array_filter($listeners, static function(RegisteredListener $listener) : bool{ return $listener instanceof RegisteredAsyncListener; });
-				$asyncListenersByPriority[$priority] = array_merge($asyncListenersByPriority[$priority] ?? [], $asyncListeners);
+			foreach($currentList->handlerSlots as $priority => $listenersToSort){
+				foreach($listenersToSort as $listener){
+					if(!$listener instanceof RegisteredAsyncListener){
+						$listeners[$priority][] = $listener;
+					}elseif(!$listener->canBeCalledConcurrently()){
+						$asyncListeners[$priority][] = $listener;
+					}else{
+						$exclusiveAsyncListeners[$priority][] = $listener;
+					}
+				}
 			}
 		}
-		foreach($asyncListenersByPriority as $priority => $asyncListeners){
-			usort($asyncListeners, static function(RegisteredAsyncListener $a, RegisteredAsyncListener $b) : int{
-				// concurrent listeners are sorted to the end of the list
-				return $b->canBeCalledConcurrently() <=> $a->canBeCalledConcurrently();
-			});
-			$listenersByPriority[$priority] = array_merge($listenersByPriority[$priority] ?? [], $asyncListeners);
-		}
+		$listenersByPriority = array_merge_recursive($listeners, $asyncListeners, $exclusiveAsyncListeners);
```

Comment:
Aren't exclusive async listeners supposed to be run before normal async listeners?

Reply:
That's the debate going on in my neurons. I don't have any pros or cons, really.

Reply:
The defining feature of exclusive listeners is that no other listener can be invoked before the exclusive listener's promise is resolved. The ordering isn't really significant.

Reply:
Looks like this function isn't actually used for async events anyway.